### PR TITLE
MPI_T general docs

### LINKED ISSUE #7519: Document some Open MPI-specific MPI_T behavior
Need to document the following:

* Values set via MPI_T are the highest precedence (e.g., even higher that values set by CLI).
* If you set framework values (e.g., set `btl` before `MPI_INIT`, it'll actually affect component selection).
* We have a CVAR per MCA framework that will tell you which component(s) are currently selected / available (after `MPI_INIT`).
* After `MPI_INIT`, most CVARs become read-only.
* There is no CVAR that lists all the frameworks.
* Groups are related projects, frameworks, components.
* MCA variables *are* CVARs.
* As of this version, very few PVARs.

This can probably go in a man page somewhere, and/or the FAQ...?

FYI @hjelmn

### PR DESCRIPTION
**Pull request by jsquyres**
Document Open MPI's MPI_T implementation. Also do it in Markdown rather than nroff -- it's so much easier to read/write.

Fixes #7519

### CODE REVIEW

**Comment by jsquyres** [comment]
Still need to create infrastructure to pandoc .X.md to .X files.

**Comment by jsquyres** [comment]
Still to do:

* Do more testing with converting the `.3` man pages to `.md` (I converted a few just for testing, but haven't integrated them in yet).
* Determine how we want to render the `.md` files on the Open MPI web site.
    * This affects how we link to man pages from each other.
* Convert all existing Open MPI nroff pages to `.md`.  Pandoc itself can read in nroff and emit Markdown, which gives a very good starting point for the Markdown.

File: configure.ac [hunk 1]
```diff
@@ -70,12 +70,46 @@ OPAL_LOAD_PLATFORM
...
+dnl Intentionally not putting an AC HELP_STRING here so that this option
+dnl does not show up in "./configure --help" output.
+AC_ARG_ENABLE([dist])
+
+AC_PATH_PROG([PYTHON], [python])
+
+AC_PATH_PROG([PANDOC], [pandoc])
+
+happy=0
+AS_IF([test -n "$PANDOC"],
```
Comment (bwbarrett) [inline configure.ac:90]: Why do we need --enable-dist again?  --enable-dist is *EVIL*; it makes "make distcheck" not work properly.  Much better would be to just abort during "make dist" if the pandoc string is the "fake" pandoc.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/38ff117fdf638fc4d11af6108c457a6a

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/48a4b8fb265ee172fd9f77a678fced1f

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0512939fd89b5a6fdaaabb6f59d3fc3b

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/06b8b6651b68263faae0903eece7c544

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8fccee18517b3102f77a939218ccb1fc

**Comment by jsquyres** [comment]
@bwbarrett You got your wish -- `pandoc` is now like `flex`, meaning that the CI AMI's need to be updated to have pandoc >= v1.19.  😉

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/da673350c52ecde1c05068454c78acf5

**Comment by jsquyres** [comment]
@jjhursey Can you install pandoc >=v1.19 on your CI systems?

Thanks!

**Comment by jjhursey** [comment]
@jsquyres Do you really need pandoc 1.19 or better? Centos 7.7 comes with 1.12.3.1 from epel.

**Comment by jsquyres** [comment]
@jjhursey I have only tested:

* pandoc 1.9 that comes with RHEL 6: does not work
* pandoc 1.19: that comes with Ubuntu 18.04: works

I have not tried in between that.  I can download CentOS 7.7 and try the pandoc that comes with that, but it will take me a little bit (there's no pandoc binary on https://github.com/jgm/pandoc/releases/tag/1.12.3, unfortunately, and building pandoc is... tricky).

**Comment by bwbarrett** [comment]
I can try CentOS 7.7 sometime today.

File: configure.ac [hunk 2]
```diff
@@ -981,6 +982,42 @@ if test -z "$LEX" || \
...
+      [pandoc_version=`pandoc --version | head -n 1 | awk '{ print $2 }'`
+       pandoc_major=`echo $pandoc_version | cut -d\. -f1`
+       pandoc_minor=`echo $pandoc_version | cut -d\. -f2`
+       AC_MSG_CHECKING([pandoc version])
+       AC_MSG_RESULT([major: $pandoc_major, minor: $pandoc_minor])
+
+       AC_MSG_CHECKING([if pandoc version is >=v$min_major_version.$min_minor_version])
+       AS_IF([test $pandoc_major -lt $min_major_version], [PANDOC=])
+       AS_IF([test $pandoc_major -eq $min_major_version && test $pandoc_minor -lt 19],
```
Comment (jjhursey) [inline configure.ac:1004]:
```suggestion
       AS_IF([test $pandoc_major -eq $min_major_version && test $pandoc_minor -lt $min_minor_version],
```

**Comment by jjhursey** [comment]
I ran with `CentOS Linux release 7.7.1908 (AltArch)` and pandoc `1.12.3.1` (which is what is provided by EPEL). The man pages look like they rendered correctly to me.

So we should be able to move the minimum version down to 1.12.

File: HACKING [hunk 3]
```diff
@@ -247,3 +247,26 @@ If you do not have Flex installed, it can be downloaded from the
 following URL:
 
     https://github.com/westes/flex
+
+Use of Pandoc
+=============
+
+Similar to prior sections, tou need to read/care about this section
```
Comment (jjhursey) [inline HACKING:254]:
```suggestion
Similar to prior sections, you need to read/care about this section
```

File: configure.ac [hunk 4]
```diff
@@ -981,6 +982,42 @@ if test -z "$LEX" || \
...
+AC_PATH_PROG([PANDOC], [pandoc])
+
+# If we found Pandoc, check its version.  We need >=v1.19.
+# To be clear: I know that v1.19 works, and I know that v1.9 does not
+# work.  I did not test the versions in between to know exactly what
+# the lowest version is that works.  Someone is free to update this
+# check someday to be more accurate if they wish.
+min_major_version=1
+min_minor_version=19
+AS_IF([test -n "$PANDOC"],
```
Comment (jjhursey) [inline configure.ac:995]: It's our intention that users do not need pandoc to build OMPI, but with this configure check they will need to have it installed (right?) to configure/make/make-install OMPI from a tarball. Is there a way to make this check conditional if the man pages were created?

File: ompi/mpi/man/man5/MPI_T.5.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
+# NAME
+
+Open MPI's MPI_T interface - General information
+
+# DESCRIPTION
+
+There are a few Open MPI-specific notes worth mention about its `MPI_T` interface implementation.
```
Comment (jjhursey) [inline ompi/mpi/man/man5/MPI_T.5.md:7]:
```suggestion
There are a few Open MPI-specific notes worth mentioning about its `MPI_T` interface implementation.
```

**Comment by jjhursey** [comment]
I adjusted the IBM CI Centos 7.7 image to include the following. It's passing now 🎺 🎊 .
```
yum -y install epel-release
yum -y install pandoc man
```
That'll pick up pandoc `pandoc 1.12.3.1`

File: configure.ac (hunk 4 above)
Comment (jsquyres) [reply configure.ac:995]: FWIW, we do the same thing for flex (i.e., we always check for it, but we only error if a file that will be generated by flex is not already present -- i.e., if we're in a distribution tarball that already contains the generated file, then we don't need flex).

**Comment by jsquyres** [comment]
@artemry-mlnx Can you please install pandoc in your CI image?  Any version of pandoc >= v1.12 will be fine.

File: configure.ac (hunk 4 above)
Comment (jjhursey) [reply configure.ac:995]: Ok. I wasn't sure how that was handled. It read like the user would need pandoc even when using a distribution tarball.

**Comment by artemry-mlnx** [comment]
@jsquyres 
Sure, will update it shortly.

**Comment by jsquyres** [comment]
@artemry-mlnx It didn't find pandoc for some reason:

```
checking for pandoc... no
configure: WARNING: *** Could not find a suitable pandoc on your system.
configure: WARNING: *** You need pandoc >=1.12 to build Open MPI man pages.
configure: WARNING: *** See pandoc.org.
configure: WARNING: *** YOU DO NOT NEED PANDOC WHEN BUILDING DISTRIBUTION TARBALLS!
configure: error: Cannot continue
```

**Comment by artemry-mlnx** [comment]
There were some issues with the CI docker container - fixed, it should work now.

**Comment by jsquyres** [comment]
@artemry-mlnx Looks like it still didn't fix the pandoc problem.

File: ompi/Makefile.am [hunk 6]
```diff
@@ -172,6 +175,8 @@ ompidir = $(ompiincludedir)/$(subdir)
 nobase_ompi_HEADERS = $(headers)
 endif
 
+include $(top_srcdir)/Makefile.ompi-rules
```
Comment (bwbarrett) [inline ompi/Makefile.am:178]: why?

File: config/opal_setup_man_pages.m4 [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                 [AC_MSG_RESULT([yes])],
+                 [AC_MSG_RESULT([no])])
+          ])
+
+    AS_IF([test -z "$PANDOC" || test -n "`echo $PANDOC | $GREP missing`"],
+          [AS_IF([test ! -f "$srcdir/ompi/mpi/man/man5/MPI_T.5"],
+                 [AC_MSG_WARN([*** Could not find a suitable pandoc on your system.])
+                  AC_MSG_WARN([*** You need pandoc >=$min_major_version.$min_minor_version to build Open MPI man pages.])
+                  AC_MSG_WARN([*** See pandoc.org.])
+                  AC_MSG_WARN([*** YOU DO NOT NEED PANDOC WHEN BUILDING DISTRIBUTION TARBALLS!])
```
[the remark this answers is not in the document]
Comment (jsquyres) [reply config/opal_setup_man_pages.m4:72]: I copied the language from where we stated this about flex: https://github.com/open-mpi/ompi/blob/master/configure.ac#L979

I didn't find this ambiguous, but I have no strong feelings here: do you have a suggestion for different text?  The goal was to have a short, simple statement that end users would read and be able to understand.  Perhaps:

IF YOU ARE BUILDING A TARBALL DOWNLOADED FROM WWW.OPEN-MPI.ORG, YOU DO NOT NEED PANDOC!

?

File: configure.ac [hunk 8]
```diff
@@ -70,13 +70,14 @@ OPAL_LOAD_PLATFORM
 # Start it up
 #
 
+OPAL_CONFIGURE_SETUP
```
[the remark this answers is not in the document]
Comment (jsquyres) [reply configure.ac:73]: I noticed that this was actually in the wrong order while working on this (i.e., Perl was checked before the subtitle instead of the other way around).

I can separate it out into a separate commit, if you prefer.

File: ompi/Makefile.am (hunk 6 above)
Comment (jsquyres) [reply ompi/Makefile.am:178]: I think I mentioned this one verbally on the phone: I changed the build scheme to traverse into the two man page directories so that it is easier for a developer writing man pages to iterate over writing a man page and checking how it rendered.

Specifically, when there is a `Makefile.am` in the man page directories, you can just invoke `make` in an individual directory and see how the page you are working on rendered.  Otherwise, you have to invoke `make` at the top-level `ompi` dir, which has to traverse a bunch of directories before it renders the page you are working on.

Put simply: it's an optimization for those who are working on man pages.  I figured the trivial time that it adds to the general build was worth it.

Comment (jsquyres) [reply ompi/Makefile.am:178]: Wait; I read Brian's comment wrong / out of context.

I think that this `include` is left over from when I was experimenting with a few different ways to get the makefile rules in the right places.  I'll double check, but I think this new `include` here in `ompi/Makefile.am` should be deleted.

File: config/opal_setup_man_pages.m4 (hunk 7 above)
[the remark this answers is not in the document]
Comment (bwbarrett) [reply config/opal_setup_man_pages.m4:72]: What about "If you are building from a tarball downloaded from www.open-mpi.org, you do not need pandoc?"

File: configure.ac (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwbarrett) [reply configure.ac:73]: no, fair enough.

File: ompi/Makefile.am (hunk 6 above)
Comment (bwbarrett) [reply ompi/Makefile.am:178]: Yes, I was just asking about the include of the Makefile.ompi-rules, which I don't think is needed when the manpages are split out.

File: configure.ac (hunk 8 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply configure.ac:73]: Yeah, but now I can't un-see it.  Fix inbound.

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by bwbarrett** [comment]
bot:aws:retest

**Comment by bwbarrett** [comment]
bot:aws:retest

**Comment by bwbarrett** [comment]
bot:aws:retest

**Comment by bwbarrett** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
@bwbarrett Can you remove your negative review?

**Comment by bwbarrett** [comment]
@jsquyres sorry, I thought I had approved last night :(.
